README.md
# Presto client library for Ruby

[![Build Status](https://travis-ci.org/treasure-data/presto-client-ruby.svg?branch=master)](https://travis-ci.org/treasure-data/presto-client-ruby)

Presto is a distributed SQL query engine for big data:
https://prestosql.io/

This is a client library for Ruby to run queries on Presto.

## Example

```ruby
require 'presto-client'

# create a client object:
client = Presto::Client.new(
  server: "localhost:8880",   # required option
  ssl: {verify: false},
  catalog: "native",
  schema: "default",
  user: "frsyuki",
  password: "********",
  time_zone: "US/Pacific",
  language: "English",
  properties: {
    "hive.force_local_scheduling": true,
    "raptor.reader_stream_buffer_size": "32MB"
  },
  http_proxy: "proxy.example.com:8080",
  http_debug: true,
)

# run a query and get results as an array of arrays:
columns, rows = client.run("select * from sys.node")
rows.each {|row|
  p row  # row is an array
}

# run a query and get results as an array of hashes:
results = client.run_with_names("select alpha, 1 AS beta from tablename")
results.each {|row|
  p row['alpha']   # access by name
  p row['beta']
  p row.values[0]  # access by index
  p row.values[1]
}

# run a query and fetch results streamingly:
client.query("select * from sys.node") do |q|
  # get columns:
  q.columns.each {|column|
    puts "column: #{column.name}.#{column.type}"
  }

  # get query results. it feeds more rows until
  # query execution finishes:
  q.each_row {|row|
    p row  # row is an array
  }
end

# killing a query
query = client.query("select * from sys.node")
query_id = query.query_info.query_id
query.each_row {|row| ... }  # when a thread is processing the query,
client.kill(query_id)  # another thread / process can kill the query.
```

## Build models

```
$ bundle exec rake modelgen:latest
```

## Options

* **server** sets address (and port) of a Presto coordinator server.
* **ssl** enables https.
  * Setting `true` enables SSL and verifies server certificate using system's built-in certificates.
  * Setting `{verify: false}` enables SSL but doesn't verify server certificate.
  * Setting a Hash object enables SSL and verify server certificate with options:
    * **ca_file**: path of a CA certification file in PEM format
    * **ca_path**: path of a CA certification directory containing certifications in PEM format
    * **cert_store**: a `OpenSSL::X509::Store` object used for verification
    * **client_cert**: a `OpenSSL::X509::Certificate` object as client certificate
    * **client_key**: a `OpenSSL::PKey::RSA` or `OpenSSL::PKey::DSA` object used for client certificate
* **catalog** sets catalog (connector) name of Presto such as `hive-cdh4`, `hive-hadoop1`, etc.
* **schema** sets default schema name of Presto. You need to use qualified name like `FROM myschema.table1` to use non-default schemas.
* **source** sets source name to connect to a Presto. This name is shown on Presto web interface.
* **client_info** sets client info to queries. It can be a string to pass a raw string, or an object that can be encoded to JSON.
* **client_tags** sets client tags to queries. It needs to be an array of strings. The tags are shown on web interface.
* **user** sets user name to connect to a Presto.
* **password** sets a password to connect to Presto using basic auth.
* **time_zone** sets time zone of queries. Time zone affects some functions such as `format_datetime`.
* **language** sets language of queries. Language affects some functions such as `format_datetime`.
* **properties** set session properties. Session properties affect internal behavior such as `hive.force_local_scheduling: true`, `raptor.reader_stream_buffer_size: "32MB"`, etc.
* **query_timeout** sets timeout in seconds for the entire query execution (from the first API call until there're no more output data). If timeout happens, client raises PrestoQueryTimeoutError. Default is nil (disabled).
* **plan_timeout** sets timeout in seconds for query planning execution (from the first API call until result columns become available). If timeout happens, client raises PrestoQueryTimeoutError. Default is nil (disabled).
* **http_headers** sets custom HTTP headers. It must be a Hash of string to string.
* **http_proxy** sets host:port of a HTTP proxy server.
* **http_debug** enables debug message to STDOUT for each HTTP requests.
* **http_open_timeout** sets timeout in seconds to open new HTTP connection.
* **http_timeout** sets timeout in seconds to read data from a server.
* **model_version** set the presto version to which a job is submitted. Supported versions are 0.205, 0.178, 0.173, 0.153 and 0.149. Default is 0.205.

See [RDoc](http://www.rubydoc.info/gems/presto-client/) for the full documentation.

## Development

### Releasing a new version

1. Update lib/presto/client/version.rb
2. Update ChangeLog
3. git commit -am "vX.Y.Z"
4. git tag "vX.Y.Z"
5. git push --tags


ChangeLog
2019-08-09 version 0.6.1

* Fix WriterTarget 316 model class name

2019-07-23 version 0.6.0

* Support presto 316 model class


2019-01-30 version 0.5.14

* Added `Query#current_results_headers` that returns HTTP response headers

2018-12-10 version 0.5.13

* Added `query_timeout` and `plan_timeout` options with default disabled
* Changed timer to use CLOCK_MONOTONIC to avoid unexpected behavior when
  system clock is updated

2018-08-07 version 0.5.12

* Upgrade to Presto 0.205 model

2018-06-27 version 0.5.11

* Support multiple session properties
* Check invalid JSON data response

2018-03-22 version 0.5.10

* Added client_info, client_tags, and http_headers options.

2018-03-02 version 0.5.9

* Added error_name field at PrestoQueryError

2017-11-13 version 0.5.8

* Added `Client#kill(query_id)` method.
* Added additional checking of internal exceptions so that client doesn't
  silently return in case when Presto query is killed and Presto returns a
  valid `200 OK` response with `result_uri: null`.
* Fixed `undefined local variable 'body'` error that was possibly happening
  when Presto returned an unexpected data structure.

2017-08-28 version 0.5.7
* Support a password option with HTTP basic auth
* Changed retry timeout from hard coded 2h to configurable default 2min
* Fix too deep nested json failure

2017-07-03 version 0.5.6:
* Added missing inner class models for version 0.178

2017-06-28 version 0.5.5:
* Added support for model version 0.178

2017-05-15 version 0.5.4:

* Support "Content-Type: application/x-msgpack" for more efficient parsing of
  HTTP response body.
* Added "enable_x_msgpack: true" option to send Accept header with
  application/x-msgpack.


2017-04-26 version 0.5.3:

* Added support for model version 0.173.
* Changed the default latest model version to 0.173.
* Fixed compatibility with the new major version of Farady
* Require Faraday 0.12 or later


2017-02-01 version 0.5.2:

* Relax dependent version of Faraday to be able to use all 0.x versions.
* Fix build script that was broken due to new major version of rake.


2016-11-01 version 0.5.1:

* Assume ConnectorId as a primitive type to be able to decode "connectorId"
  fields.


2016-10-28 version 0.5.0:

* Support multiple model versions
* Added support for model version 0.153.
* Changed the default latest model version to 0.513.


2016-08-09 version 0.4.17:

* Added support for :ssl option.


2016-08-03 version 0.4.16:

* Upgraded Presto model version to 0.151


2016-08-03 version 0.4.15:

* decode method of model classes validate Hash type


2016-08-02 version 0.4.14:

* Added support for resuming fetching query results by using new `Query.resume(next_uri, options)` method (@tetrakai++)


2016-08-02 version 0.4.13:

* Added support for :http_proxy option to use a HTTP proxy server
* Added support for hashed Client response using `run_with_names` (thanks to MoovWeb for allowing me to contribute)
* Upgraded Presto model version to 0.134

2015-04-01 version 0.4.5:

* Upgraded Presto model version to 0.99


2014-11-20 version 0.4.3:

* Updated gem dependency to accept faraday ~> 0.9.x as well as ~> 0.8.8


2014-10-15 version 0.4.2:

* Added support for :properties option to set session properties introduced
since Presto 0.78


2014-06-12 version 0.4.1:

* Added EquiJoinClause model class
* Added StageId#query_id and #id methods
* Added TaskId#query_id, #stage_id and #id methods


2014-06-10 version 0.4.0:

* Added Query#current_results, #advance and #query_info for advanced users
* Generate model classes from Presto source code to include complete classes


2014-05-06 version 0.3.3:

* Added :time_zone and :language options added by Presto 0.66


2014-04-01 version 0.3.2:

* Fixed a problem that client skips the last chunk if result is large


2014-01-30 version 0.3.1:

* Added http_debug option
* Disabled HTTP debug logging by default


2014-01-22 version 0.3.0:

* Added http_timeout option
* Added http_open_timeout option
* Changed Query.start API to start(query, options) to http options


2014-01-22 version 0.2.0:

* Added Query#cancel
* Added Query#close
* Added Client#run
* Changed required_ruby_version from 1.9.3 to 1.9.1


2014-01-07 version 0.1.0:

* First release


presto-client.rb

module PrestoClient
  VERSION = "0.1.0"

  require 'faraday'
  require 'json'

  class ClientSession
    def initialize(options)
      @server = options[:server]
      @user = options[:user]
      @source = options[:source]
      @catalog = options[:catalog]
      @schema = options[:schema]
      @debug = !!options[:debug]
    end

    attr_reader :server
    attr_reader :user
    attr_reader :source
    attr_reader :catalog
    attr_reader :schema

    def debug?
      @debug
    end
  end

  #class StageStats
  #  attr_reader :stage_id
  #  attr_reader :state
  #  attr_reader :done
  #  attr_reader :nodes
  #  attr_reader :total_splits
  #  attr_reader :queued_splits
  #  attr_reader :running_splits
  #  attr_reader :completed_splits
  #  attr_reader :user_time_millis
  #  attr_reader :cpu_time_millis
  #  attr_reader :wall_time_millis
  #  attr_reader :processed_rows
  #  attr_reader :processed_bytes
  #  attr_reader :sub_stages
  #
  #  def initialize(options={})
  #    @stage_id = options[:stage_id]
  #    @state = options[:state]
  #    @done = options[:done]
  #    @nodes = options[:nodes]
  #    @total_splits = options[:total_splits]
  #    @queued_splits = options[:queued_splits]
  #    @running_splits = options[:running_splits]
  #    @completed_splits = options[:completed_splits]
  #    @user_time_millis = options[:user_time_millis]
  #    @cpu_time_millis = options[:cpu_time_millis]
  #    @wall_time_millis = options[:wall_time_millis]
  #    @processed_rows = options[:processed_rows]
  #    @processed_bytes = options[:processed_bytes]
  #    @sub_stages = options[:sub_stages]
  #  end
  #
  #  def self.decode_hash(hash)
  #    new(
  #      stage_id: hash["stageId"],
  #      state: hash["state"],
  #      done: hash["done"],
  #      nodes: hash["nodes"],
  #      total_splits: hash["totalSplits"],
  #      queued_splits: hash["queuedSplits"],
  #      running_splits: hash["runningSplits"],
  #      completed_splits: hash["completedSplits"],
  #      user_time_millis: hash["userTimeMillis"],
  #      cpu_time_millis: hash["cpuTimeMillis"],
  #      wall_time_millis: hash["wallTimeMillis"],
  #      processed_rows: hash["processedRows"],
  #      processed_bytes: hash["processedBytes"],
  #      sub_stages: hash["subStages"].map {|h| StageStats.decode_hash(h) },
  #    )
  #  end
  #end

  class StatementStats
    attr_reader :state
    attr_reader :scheduled
    attr_reader :nodes
    attr_reader :total_splits
    attr_reader :queued_splits
    attr_reader :running_splits
    attr_reader :completed_splits
    attr_reader :user_time_millis
    attr_reader :cpu_time_millis
    attr_reader :wall_time_millis
    attr_reader :processed_rows
    attr_reader :processed_bytes
    #attr_reader :root_stage

    def initialize(options={})
      @state = state
      @scheduled = scheduled
      @nodes = nodes
      @total_splits = total_splits
      @queued_splits = queued_splits
      @running_splits = running_splits
      @completed_splits = completed_splits
      @user_time_millis = user_time_millis
      @cpu_time_millis = cpu_time_millis
      @wall_time_millis = wall_time_millis
      @processed_rows = processed_rows
      @processed_bytes = processed_bytes
      #@root_stage = root_stage
    end

    def self.decode_hash(hash)
      new(
        state: hash["state"],
        scheduled: hash["scheduled"],
        nodes: hash["nodes"],
        total_splits: hash["totalSplits"],
        queued_splits: hash["queuedSplits"],
        running_splits: hash["runningSplits"],
        completed_splits: hash["completedSplits"],
        user_time_millis: hash["userTimeMillis"],
        cpu_time_millis: hash["cpuTimeMillis"],
        wall_time_millis: hash["wallTimeMillis"],
        processed_rows: hash["processedRows"],
        processed_bytes: hash["processedBytes"],
        #root_stage: StageStats.decode_hash(hash["rootStage"]),
      )
    end
  end

  class Column
    attr_reader :name
    attr_reader :type

    def initialize(options={})
      @name = options[:name]
      @type = options[:type]
    end

    def self.decode_hash(hash)
      new(
        name: hash["name"],
        type: hash["type"],
      )
    end
  end

  class QueryResults
    attr_reader :id
    attr_reader :info_uri
    attr_reader :partial_cache_uri
    attr_reader :next_uri
    attr_reader :columns
    attr_reader :data
    attr_reader :stats
    attr_reader :error

    def initialize(options={})
      @id = options[:id]
      @info_uri = options[:info_uri]
      @partial_cache_uri = options[:partial_cache_uri]
      @next_uri = options[:next_uri]
      @columns = options[:columns]
      @data = options[:data]
      @stats = options[:stats]
      @error = options[:error]
    end

    def self.decode_hash(hash)
      new(
        id: hash["id"],
        info_uri: hash["infoUri"],
        partial_cache_uri: hash["partialCancelUri"],
        next_uri: hash["nextUri"],
        columns: hash["columns"] ? hash["columns"].map {|h| Column.decode_hash(h) } : nil,
        data: hash["data"],
        stats: StatementStats.decode_hash(hash["stats"]),
        error: hash["error"]  # TODO
      )
    end
  end

  module PrestoHeaders
    PRESTO_USER = "X-Presto-User"
    PRESTO_SOURCE = "X-Presto-Source"
    PRESTO_CATALOG = "X-Presto-Catalog"
    PRESTO_SCHEMA = "X-Presto-Schema"

    PRESTO_CURRENT_STATE = "X-Presto-Current-State"
    PRESTO_MAX_WAIT = "X-Presto-Max-Wait"
    PRESTO_MAX_SIZE = "X-Presto-Max-Size"
    PRESTO_PAGE_SEQUENCE_ID = "X-Presto-Page-Sequence-Id"
  end

  class StatementClient
    HEADERS = {
      "User-Agent" => "presto-ruby/#{VERSION}"
    }

    def initialize(faraday, session, query)
      @faraday = faraday
      @faraday.headers.merge!(HEADERS)

      @session = session
      @query = query
      @closed = false
      @exception = nil
      post_query_request!
    end

    def post_query_request!
      response = @faraday.post do |req|
        req.url "/v1/statement"

        if v = @session.user
          req.headers[PrestoHeaders::PRESTO_USER] = v
        end
        if v = @session.source
          req.headers[PrestoHeaders::PRESTO_SOURCE] = v
        end
        if catalog = @session.catalog
          req.headers[PrestoHeaders::PRESTO_CATALOG] = catalog
        end
        if v = @session.schema
          req.headers[PrestoHeaders::PRESTO_SCHEMA] = v
        end

        req.body = @query
      end

      # TODO error handling
      if response.status != 200
        raise "Failed to start query: #{response.body}"  # TODO error class
      end

      body = response.body
      hash = JSON.parse(body)
      @results = QueryResults.decode_hash(hash)
    end

    private :post_query_request!

    attr_reader :query

    def debug?
      @session.debug?
    end

    def closed?
      @closed
    end

    attr_reader :exception

    def exception?
      @exception
    end

    def query_failed?
      @results.error != nil
    end

    def query_succeeded?
      @results.error == nil && !@exception && !@closed
    end

    def current_results
      @results
    end

    def has_next?
      !!@results.next_uri
    end

    def advance
      if closed? || !has_next?
        return false
      end
      uri = @results.next_uri

      start = Time.now
      attempts = 0

      begin
        begin
          response = @faraday.get do |req|
            req.url uri
          end
        rescue => e
          @exception = e
          raise @exception
        end

        if response.status == 200 && !response.body.to_s.empty?
          @results = QueryResults.decode_hash(JSON.parse(response.body))
          return true
        end

        if response.status != 503  # retry on 503 Service Unavailable
          # deterministic error
          @exception = StandardError.new("Error fetching next at #{uri} returned #{response.status}: #{response.body}")  # TODO error class
          raise @exception
        end

        attempts += 1
        sleep attempts * 0.1
      end while (Time.now - start) < 2*60*60 && !@closed

      @exception = StandardError.new("Error fetching next")  # TODO error class
      raise @exception
    end

    def close
      return if @closed

      # cancel running statement
      if uri = @results.next_uri
        # TODO error handling
        # TODO make async reqeust and ignore response
        @faraday.delete do |req|
          req.url uri
        end
      end

      @closed = true
      nil
    end
  end

  class Query
    def self.start(session, query)
      faraday = Faraday.new(url: "http://#{session.server}") do |faraday|
        faraday.request :url_encoded
        faraday.response :logger
        faraday.adapter Faraday.default_adapter
      end

      new StatementClient.new(faraday, session, query)
    end

    def initialize(client)
      @client = client
    end

    def wait_for_data
      while @client.has_next? && @client.current_results.data == nil
        @client.advance
      end
    end

    private :wait_for_data

    def columns
      wait_for_data

      raise_error unless @client.query_succeeded?

      return @client.current_results.columns
    end

    def each_row(&block)
      wait_for_data

      raise_error unless @client.query_succeeded?

      if self.columns == nil
        raise "Query #{@client.current_results.id} has no columns"
      end

      begin
        if data = @client.current_results.data
          data.each(&block)
        end
        @client.advance
      end while @client.has_next?
    end

    def raise_error
      if @client.closed?
        raise "Query aborted by user"
      elsif @client.exception?
        raise "Query is gone: #{@client.exception}"
      elsif @client.query_failed?
        results = @client.current_results
        # TODO error location
        raise "Query #{results.id} failed: #{results.error}"
      end
    end

    private :raise_error
  end

  class Client
    def initialize(options)
      @session = ClientSession.new(options)
    end

    def query(query)
      Query.start(@session, query)
    end
  end
end

require 'pp'

client = PrestoClient::Client.new(
  server: "localhost:8880",
  user: "frsyuki",
  catalog: "native",
  schema: "default",
  debug: true
)

q = client.query("select * from sys.query")
p q.columns
q.each_row {|row|
  p row
}

